src/main/java/frc/team2412/robot/util/PFFController.java

```diff
@@ -0,0 +1,196 @@
+package frc.team2412.robot.util;
+
+import edu.wpi.first.wpilibj.Timer;
+import org.frcteam2910.common.math.MathUtils;
+import org.frcteam2910.common.math.Vector2;
+import org.frcteam2910.common.math.Vector3;
+
+/**
+ * Class for PFF controllers that provide P controllers but with the advantage of FeedForward
+ *
+ * @author Alex Stedman
+ * @param <T>
+ *            type of input/output
+ */
+@SuppressWarnings("unused")
+public class PFFController<T> {
+    private final double p, f;
+    private final Timer timer;
+    private final PFFOperator<T> op;
+    private double targetPositionTolerance = 0.1;
+
+    /**
+     * functional interface for core operator used by the controller
+     *
+     * @author Alex Stedman
+     * @param <T>
+     *            type of input/output
+     */
+    public interface PFFOperator<T> {
+
+        T add(T first, T second);
+
+        default T subtract(T first, T second) {
+            return add(first, scale(second, -1));
+        }
+
+        T scale(T first, double scale);
+
+        boolean equals(T first, T second, double tolerance);
+
+        T zero();
+
+        default T operate(T first, T second, double scale, double tolerance) {
+            if (equals(first, second, tolerance))
+                return zero();
+            return scale(subtract(first, second), scale);
+        }
+    }
+
+    protected PFFController(double pValue, double fValue, PFFOperator<T> operator) {
+        p = pValue;
+        f = fValue;
+        op = operator;
+        timer = new Timer();
+    }
+
+    private T targetPosition;
+
+    /**
+     * Set the target position of the controller
+     *
+     * @param target
+     *            new target position
+     * @return this
+     */
+    public PFFController<T> setTargetPosition(T target) {
+        targetPosition = target;
+        timer.reset();
+        return this;
+    }
+
+    public PFFController<T> setTargetPositionTolerance(double tolerance) {
+        targetPositionTolerance = tolerance;
+        return this;
+    }
+
+    private T pastValue;
+
+    /**
+     * output the adjustment offered by the controller
+     *
+     * @param measuredValue
+     *            measured value at current time
+     * @return new value to set
+     */
+    public T update(T measuredValue) {
+        if (pastValue == null)
+            pastValue = measuredValue;
+        if (targetPosition == null)
+            return null;
+        T velocity = op.operate(measuredValue, pastValue, f, targetPositionTolerance);
```

Comment:
I'm not sure what you're doing with `pastValue` here - measuring previous changes in value is not something normally found in feedforward control.

Reply:
It's calculating a unitless acceleration (difference in speed) or speed (difference in position).  Ideally it'd  instead use an actual acceleration or velocity value, similar to the HolonomicDriveController's `calculate` method (see [here](https://github.com/wpilibsuite/allwpilib/blob/181723e573d36048b8f99153d5f5e75388f702c6/wpimath/src/main/native/cpp/controller/HolonomicDriveController.cpp#L34), with the method signature below)

  /**
   * Returns the next output of the holonomic drive controller.
   *
   * @param currentPose The current pose.
   * @param poseRef The desired pose.
   * @param linearVelocityRefMeters The linear velocity reference.
   * @param angleRef The angular reference.
   * @return The next output of the holonomic drive controller.
   */
  @SuppressWarnings("LocalVariableName")
  public ChassisSpeeds calculate(
      Pose2d currentPose, Pose2d poseRef, double linearVelocityRefMeters, Rotation2d angleRef) {